src/dataflow/src/render/reduce.rs

```diff
@@ -280,13 +280,43 @@ where
         // update costs on relatively small updates.
         if hierarchical {
             if monotonic && is_min_or_max(&func) {
-                use differential_dataflow::operators::iterate::Variable;
-                let delay = std::time::Duration::from_nanos(10_000_000_000);
-                let retractions = Variable::new(&mut partial.scope(), delay.as_millis() as u64);
-                let thinned = partial.concat(&retractions.negate());
-                let result = build_hierarchical(thinned, &func);
-                retractions.set(&partial.concat(&result.negate()));
-                partial = result
+                // At this point, we assert that inputs are never retracted.
```

Comment:
just so I understand -- whats the failure mode here if we screw up the monotonicity analysis and some records are retracted? my understanding is that it would only matter for retractions of the current max / min for a key and we would ignore the retraction and continue to output the now retracted min/max -- is that right?

Reply:
I believe we just fail to retract the record. This has some negative implications if there are multiple aggregates and others are successfully retracted (e.g. a sum) because the merging step will likely panic with a complaint about missing aggregate values.

Reply:
Actually, it seems like there could have been a bug here! I added a `consolidate` because the stream that comes in could be "logically monotonic" but contain retractions (e.g. x - (x - y) for monotonic y) and we weren't zeroing out the cancelations before promoting things to monoids.

Reply:
that makes a lot of sense / its scary how easy it is to mess that up